src/ray/gcs/gcs_server/ray_sync.h

```diff
@@ -0,0 +1,114 @@
+
+  void Start() {
+    poller_->Start();
+    broadcast_thread_ = std::make_unique<std::thread>([this]() {
+      SetThreadName("resource_report_broadcaster");
+      boost::asio::io_service::work work(broadcast_service_);
+      broadcast_service_.run();
+    });
+    ticker_.RunFnPeriodically(
+        [this] {
+          auto beg = resources_buffer_.begin();
+          auto ptr = beg;
+          static auto max_batch = RayConfig::instance().resource_broadcast_batch_size();
+          for (size_t cnt = resources_buffer_proto_.batch().size();
+               cnt < max_batch && cnt < resources_buffer_.size(); ++ptr, ++cnt) {
+            resources_buffer_proto_.add_batch()->mutable_data()->Swap(&ptr->second);
+          }
+          resources_buffer_.erase(beg, ptr);
+          broadcaster_->SendBroadcast(std::move(resources_buffer_proto_));
+          resources_buffer_proto_.Clear();
+        },
+        RayConfig::instance().raylet_report_resources_period_milliseconds(),
+        "RaySyncer.deadline_timer.report_resource_report");
+  }
+
+  void Stop() {
+    poller_->Stop();
+    if (broadcast_thread_ != nullptr) {
+      broadcast_service_.stop();
+      if (broadcast_thread_->joinable()) {
+        broadcast_thread_->join();
+      }
+    }
+  }
+
+  // External API
+  template <typename T>
+  void Update(T update) {
+    static_assert(std::is_same_v<T, rpc::NodeResourceChange> ||
+                      std::is_same_v<T, rpc::ResourcesData>,
+                  "unknown type");
+
+    if constexpr (std::is_same_v<T, rpc::NodeResourceChange>) {
+      resources_buffer_proto_.add_batch()->mutable_change()->Swap(&update);
+    } else if constexpr (std::is_same_v<T, rpc::ResourcesData>) {
+      if (update.should_global_gc() || update.resources_total_size() > 0 ||
+          update.resources_available_changed() || update.resource_load_changed()) {
+        update.clear_resource_load();
+        update.clear_resource_load_by_shape();
+        update.clear_resources_normal_task();
+        auto &orig = resources_buffer_[update.node_id()];
+        orig.Swap(&update);
+      }
+    }
+  }
+
+  void AddNode(const rpc::GcsNodeInfo &node_info) {
+    broadcaster_->HandleNodeAdded(node_info);
+    poller_->HandleNodeAdded(node_info);
+  }
+
+  void RemoveNode(const rpc::GcsNodeInfo &node_info) {
+    broadcaster_->HandleNodeRemoved(node_info);
+    poller_->HandleNodeRemoved(node_info);
+    resources_buffer_.erase(node_info.node_id());
+  }
+
+  std::string DebugString() { return broadcaster_->DebugString(); }
+
+ private:
+  PeriodicalRunner ticker_;
+  rpc::ResourceUsageBroadcastData resources_buffer_proto_;
+  std::unique_ptr<::ray::gcs::GrpcBasedResourceBroadcaster> broadcaster_;
+  std::unique_ptr<::ray::gcs::GcsResourceReportPoller> poller_;
+  absl::flat_hash_map<std::string, rpc::ResourcesData> resources_buffer_;
+  std::unique_ptr<std::thread> broadcast_thread_;
+  instrumented_io_context broadcast_service_;
```

Comment:
Who is using these? I didn't find the usage.

Reply:
Good catch!